lib/modules.nix

```diff
@@ -153,8 +154,74 @@ rec {
             # a `_module.args.pkgs = import (fetchTarball { ... }) {}` won't
             # start a download when `pkgs` wasn't evaluated.
             type = types.lazyAttrsOf types.raw;
-            internal = true;
-            description = "Arguments passed to each module.";
+            # Only render documentation once at the root of the option tree,
+            # not for all individual submodules.
+            internal = prefix != [];
+            example = literalExpression ''
+              {
+                unstable = import nixpkgs {};
+                myLib = { addOne = x: x + 1; };
```

Comment:
This is actually an antipattern. It works for "split configuration" but is a bad idea for reusable modules.
It's better than `lib.extend` by virtue of being slightly more composable, but still vulnerable to name collisions.
The most robust method of using "libraries" is to import them directly into the lexical scope: `let myLib = import ./myLlib; in .....`. The root values in file imports _are cached_, so there's no loss of performance. Function invocations are not, so we need an alternative place to memoize those and internal options fit that purpose, without the risk of name collisions or organic accumulation of "utils" junk libraries.

Reply:
Do you have suggestions for better examples? Or should we drop examples completely?

Reply:
I honestly don't know of a good example. The ones in nixpkgs itself aren't actually good either. They're highly specific and/or workarounds.

`unstable` is the best example I know of, only slightly better than `myLib`. I'd call it `pkgsUnstable` though.

I think we need to clarify what are the "acceptable" use cases:
 - end user configurations that are not intended to be included into anything that isn't defined "locally". Here convenience wins over reusability, but it does mean turning configs into modules will be harder.
 - specific features of configuration systems, maybe?
    - for example NixOS defines `noUserModules` to solve quite a NixOS-specific problem (maybe I should have used an internal `specialisations` option instead? I guess that namespace was taken by an `attrsOf`. Is that sufficient reason?
    - is `baseModules` any better?

I'm really trying to like it, but not succeeding.

Reply:
I removed the example altogether for now. We can still add one later if we have a good one.